### PR TITLE
Add support for modifying direct resources of const tables

### PR DESCRIPTION
**Pull request by konne-google**

### CODE REVIEW

**Comment by jafingerhut** [comment]
The changes look reasonable to me.

It might be helpful to have an example of a request that modifies the direct resources of such a table.

In particular, it seems that such a MODIFY message must contain the field match criteria, action, and action parameters.

Should it be an error if the field match criteria do not match any of the entries of the table?

Should it be an error if the field match criteria do match an entry of the table, but the action and/or action parameters do not match the action in the table?

**Comment by jafingerhut** [comment]
Thinking a bit more about my previous comment, I guess MODIFY should already fail for any table if the field match criteria does not correspond to an entry in the table, regardless of whether it is const entries or not?

If so, that might be good to remind a read here explicitly.

It does seem "new" for const entries tables that it is also an error if the action and/or action parameters are different than what is already in the entry.

**Comment by konne-google** [comment]
jafingerhut, I clarified which match/action fields need to be specified.

**Comment by jafingerhut** [comment]
Looks like a good addition.  Thanks.  Should it go slightly further and specify that the server must return an error code of a particular name if the client does include an action or action parameters in the MODIFY request?

**Comment by konne-google** [comment]
Andy, I think it's already specified what the switch should do, namely, if you try to modify anything but the direct resources, you'll get a PERMISSION_DENIED_ERROR. I've added a sentence to call that out.

**Comment by jafingerhut** [comment]
Konstantin, sorry, I do not know how I could have missed that.  Thanks for the additional sentence.

File: docs/v1/P4Runtime-Spec.mdk [hunk 1]
```diff
@@ -2750,16 +2750,23 @@ hope that it would simplify the implementation of the P4Runtime service.
...
-the server must return a `PERMISSION_DENIED` error. However, the contents of
-such tables can be queried by the client through a `ReadRequest`. When reading
-static (immutable) entries from a constant table, the following fields --- and
-only these fields ---, must be set by the server: `table_id`, `match`, `action`,
-`priority` and `is_default_action`. In particular, we assume that constant
-tables cannot be assigned direct resources and idle timeout is not supported for
+are identified by the `is_const_table` flag in P4Info.
+
+The only write updates which are allowed for constant tables are `MODIFY`
+operations on direct resources; and the default action, assuming the default
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk]: semi-colon?

Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: use parentheses instead

**Comment by konne-google** [comment]
> The following sentence in section 9.1.2 must be updated:
> 
> > The TableEntry action field must be set for every INSERT and MODIFY update, except when resetting the default entry.
> 
> We agreed (because of static entries) that `action` would no longer be required for MODIFY. So the sentence should read something like this:
> 
> > The TableEntry action field must be set for every INSERT update but may be left unset for MODIFY updates, in which case the action specification for the table entry will not be modified (if a MODIFY update has an unset action field and does not modify any direct resource attached to the table then the MODIFY update is a no-op).
> 
> Later on in the same section, this sentence should also be updated:
> 
> > If the action field is not set (and if is_default_action is false) or if the oneof does not match the table description in the P4Info (e.g. the oneof is action_profile_member_id for a direct table), the server must return an INVALID_ARGUMENT error code.

Thanks for spotting these, and the suggested updates!

**Comment by antoninbas** [comment]
Fixes #232

**Pull request merged**